app/controllers/backoffice/providers/steps_controller.rb

```diff
@@ -0,0 +1,203 @@
+    contacts: [
+      main_contact_attributes: %i[id first_name last_name email country_phone_code phone position],
+      public_contacts_attributes: %i[id first_name last_name email country_phone_code phone position _destroy]
+    ],
+    manager: [data_administrators_attributes: %i[id first_name last_name email _destroy]]
+  }.freeze
+
+  def show
+    validate_wizard_action
+    step_state(params[:id].to_sym)
+  end
+
+  def update
+    set_completion_level if session[:wizard_action] == "create"
+    provider_id = session[:provider_id]
+    saved_params = session[provider_id]
+    provider_attrs = saved_params.merge permitted_step_attributes
+
+    @provider = Provider.new(provider_attrs)
+    if @provider.valid?
+      session[provider_id] = provider_attrs
+      redirect_to_next_step(params[:commit])
+    else
+      render :show
+    end
+  end
+
+  private
+
+  def target_step(commit)
+    if commit == next_title
+      return next_step
+    elsif commit == back_title
+      return prev_step
+    else
+      target_step_value = commit.split[-1].downcase.to_sym
+      return target_step_value if steps.include? target_step_value
+    end
+    raise CommitValueError, "Unknown commit value"
+  end
+
+  def set_completion_level
+    completion_level = [current_step_index, session[:wizard_completion_level]].max
+    session[:wizard_completion_level] = completion_level
+  end
+
+  def redirect_to_next_step(commit)
+    if commit == submit_title
+      finish_wizard_path
+    else
+      target_step_value = target_step(commit)
+      step_state(target_step_value.to_sym)
+      render turbo_stream:
+               turbo_stream.update(
+                 "provider-form",
+                 partial: partial_path(target_step_value),
+                 locals: {
+                   provider: @provider
+                 }
+               )
+    end
+  end
+
+  def clear_session_data
+    session.delete session[:provider_id]
+    session.delete(:provider_id)
+    session.delete(:wizard_action)
+    session.delete(:wizard_completion_level)
+  end
+
+  def finish_wizard_path
+    provider_id = session[:provider_id]
+    saved_params = session[provider_id]
+    clear_session_data
+
+    if session[:wizard_action] == "create"
+      @provider = Provider.new(permitted_attributes(Provider).merge(saved_params))
+      if valid_model_and_urls? && @provider.save(validate: false)
+        redirect_to backoffice_provider_path(@provider), notice: "New provider created successfully"
+      else
+        render :show, status: :unprocessable_entity
+      end
+    else
+      @provider = Provider.find_by(id: provider_id)
+      @provider.update!(saved_params)
+      redirect_to backoffice_provider_path(@provider), notice: "Provider updated successfully"
+    end
+  end
+
+  def partial_path(step)
+    "backoffice/providers/steps/#{step}"
+  end
+
+  def step_state(step_to_set)
+    method("set_#{step_to_set}").call
```

Comment:
Dynamically calling methods with `method("set_#{step_to_set}").call ` is too risky (security concerns if user input is not strictly validated - its not easily maintainable even now it could be)